### PR TITLE
Harmonize BoxHandle and ArcHandle

### PR DESCRIPTION
**Pull request by scovich**
Previously `BoxHandle` was very simplistic (could only handle `Sized` types), while `ArcHandle` was overly abstracted (even a `SizedArcHandle` had type indirection that made it harder to use the resulting handle, in the common case of a rust callee who does not need to clone the handle and only needs a borrowed reference to the underlying). 

With this PR, we harmonize them into a single `Handle` concept which uses a `HandleDescriptor` trait to express four modes: {sized, unsized} x {box, arc}. The memory layout differs for sized vs. unsized types (the former being more efficient), while box vs. arc is chosen based on whether owned vs. shared reference semantics are desired. 

All handles provide an `as_ref()` method for obtaining a shared reference; arc handles provide an `as_arc()` method that returns a new Arc for the object (bumping its reference count), while box handles provide an `as_mut()` method that returns a mutable reference. In all cases, a handle remains valid until specifically dropped by the user.

### CODE REVIEW

**Review by nicklan — DISMISSED** [review]
lgtm with a few nits

File: ffi/src/handle.rs [hunk 1]
```diff
@@ -186,20 +270,18 @@ pub trait ArcHandle: Sized {
...
+/// A marker trait that allows a [Sized] type to act as its own [ArcHandle]. This is more efficient
+/// and also produces a pointer of the same type that can be used directly (no type aliasing).
+pub trait SizedArcHandle: Sized { }
 
 // A blanket implementation of `ArcHandle` for all types satisfying `SizedArcHandle`.
 //
-// Unlike the default (unsized) implementation, which must wrap the Arc in a Box in order to obtain
-// a thin pointer, the sized implementation can directly return a pointer to the Arc's underlying
+// Unlike the default (unsized) implementation, which must wrap the `Arc` in a `Box` in order to
+// obtain a thin pointer, the sized implementation can directly return a pointer to the underlying
```
[the remark this answers is not in the document]
Comment (scovich) [reply ffi/src/handle.rs:280]: Would we want both `[]`and backticks? Or just `[]`? Seems like the latter?

Also, I wonder if we _want_ links for every single mention of every class? The std rust docs don't seem overly "linky"? e.g. the doc comment for [Box::pin](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.pin) doesn't actually link to [Pin](https://doc.rust-lang.org/std/pin/struct.Pin.html), presumably because the method signature already did so. 

Anyway, I'll take a pass and be intentional about linking vs. merely backticking.

[the remark this answers is not in the document]
Comment (nicklan) [reply ffi/src/handle.rs:280]: I usually do [`Foo`] because it makes it a link and puts it in `fixed width`, which is also what the std docs do.

I agree we don't need to link _everything_, but here this was the sort of "intro" doc to the type so it seems like it would make sense to give links to the underlying types involved.

**Review by roeap — APPROVED** [review]
the fact that I at least believe I learned something from reding the docs make me fee this is a good thing :).

**Review by nicklan — COMMENTED** [review]
haven't had much time yet, but flushing the couple of comments I have so far, and will keep going tomorrow.

File: ffi/src/handle.rs [hunk 2]
```diff
@@ -1,235 +1,426 @@
...
+
+    pub trait HandleOps<T: ?Sized, M, S> {
+        type From: Sized;
+        type Raw: Sized;
+
+        fn into_handle(val: Self::From) -> NonNull<Self::Raw>;
+
+        unsafe fn as_ref(ptr: &NonNull<Self>) -> &T;
+
+        // WARNING: unimplemented when M=False
```
[the remark this answers is not in the document]
Comment (scovich) [reply ffi/src/handle.rs:139]: Not quite sure what you mean? One of the two methods is already `unimplemented!()` in each of the specific impl for `H`. 

I intentionally didn't define trait-provided methods here, so that the compiler complains if any impl misses one.

File: kernel/src/snapshot.rs [hunk 3]
```diff
@@ -109,6 +108,12 @@ pub struct Snapshot {
     schema: Schema,
 }
 
+impl Drop for Snapshot {
+    fn drop(&mut self) {
+        println!("Dropping snapshot");
+    }
+}
+
```
Comment (roeap) [inline kernel/src/snapshot.rs:116]: nit: should this use tracing - i.e. `debug!`?

**Review by roeap — COMMENTED** [review]
just did a quick pass before work - really enjoy how a lot of the `*const`, `*mut` semantics turned into the much less scary looking `Handle<...>` :).

File: kernel/src/snapshot.rs (hunk 3 above)
Comment (scovich) [reply kernel/src/snapshot.rs:116]: I wasn't actually trying to push this in the first place -- but we could do `debug!` if you think that's helpful to have in general?

File: ffi/src/handle.rs [hunk 4]
```diff
@@ -1,235 +1,417 @@
...
+    // able to impose at least the semi-sane lifetime of not outliving the `Handle` itself.
+    pub trait HandleOps<T: ?Sized, M, S> {
+        type From: Sized;
+        type Raw: Sized;
+
+        fn into_handle_ptr(val: Self::From) -> NonNull<Self::Raw>;
+
+        unsafe fn as_ref<'a>(ptr: *const Self::Raw) -> &'a T;
+
+        // WARNING: unimplemented when M=False
```
Comment (nicklan) [inline ffi/src/handle.rs:142]: Sorry, I should have just suggested the change:

```suggestion
        // WARNING: `unimplemented!()` when M=False
```

File: ffi/src/handle.rs [hunk 5]
```diff
@@ -1,235 +1,417 @@
...
+        unsafe fn as_mut<'a>(ptr: *mut Self::Raw) -> &'a mut T;
+
+        // WARNING: unimplemented when M=True
```
Comment (nicklan) [inline ffi/src/handle.rs:145]:
```suggestion
        // WARNING: `unimplemented!()` when M=True
```

File: ffi/src/lib.rs [hunk 6]
```diff
@@ -16,10 +16,40 @@ use delta_kernel::snapshot::Snapshot;
 use delta_kernel::{DeltaResult, Engine, Error};
 
 mod handle;
-use handle::{ArcHandle, SizedArcHandle, Unconstructable};
+use handle::{False, Handle, HandleDescriptor, True, Unconstructable};
 
 pub mod scan;
 
+#[repr(C)]
+pub struct Foo {
```
Comment (nicklan) [inline ffi/src/lib.rs:24]: this is test code right? we probably don't want it in here?

File: ffi/src/lib.rs [hunk 7]
```diff
@@ -404,6 +437,12 @@ struct ExternEngineVtable {
     allocate_error: AllocateErrorFn,
 }
 
+impl Drop for ExternEngineVtable {
+    fn drop(&mut self) {
+        println!("dropping engine interface");
```
Comment (nicklan) [inline ffi/src/lib.rs:442]:
```suggestion
        debug!("dropping engine interface");
```

File: ffi/src/lib.rs [hunk 8]
```diff
@@ -821,6 +864,7 @@ pub unsafe extern "C" fn visit_schema(
         }
     }
 
+    //panic!("foo");//println!("Visiting schema: {:?}", snapshot.schema());
```
Comment (nicklan) [inline ffi/src/lib.rs:867]:
```suggestion
```

File: kernel/src/snapshot.rs (hunk 3 above)
Comment (nicklan) [reply kernel/src/snapshot.rs:116]: I think tracing drops could be useful, but it's probably fine to remove for now and we can add this in if we're ever debugging a case where we think it's important.

File: ffi/src/handle.rs [hunk 9]
```diff
@@ -1,235 +1,417 @@
...
+
+    #[test]
+    fn test_handle_use_cases_compile() {
+        let f = Foo {
+            x: 10,
+            y: "hi".into(),
+        };
+        let mut h: Handle<MutableFoo> = Box::new(f).into();
+        let r = unsafe { h.as_mut() };
+        println!("{r:?}");
```
Comment (nicklan) [inline ffi/src/handle.rs:361]:
```suggestion
```

File: ffi/src/handle.rs [hunk 10]
```diff
@@ -1,235 +1,417 @@
...
+    }
+
+    pub trait Baz: Send + Sync {
+        fn squawk(&self);
+    }
+
+    impl Baz for Bar {
+        fn squawk(&self) {
+            println!("Bar!")
+        }
```
Comment (nicklan) [inline ffi/src/handle.rs:330]:
```suggestion
        fn squawk(&self) -> &str {
            "Bar!"
        }
```

File: ffi/src/handle.rs [hunk 11]
```diff
@@ -1,235 +1,417 @@
...
+        // let _ = unsafe { h.as_ref() };
+
+        let b = Bar {
+            x: 10,
+            y: "hello".into(),
+        };
+        let t: Box<dyn Baz> = Box::new(b);
+        let mut h: Handle<MutableBaz> = t.into();
+        let r = unsafe { h.as_mut() };
+        r.squawk();
```
Comment (nicklan) [inline ffi/src/handle.rs:399]:
```suggestion
        assert_eq!(r.squawk(), "Bar!");
```

File: ffi/src/handle.rs [hunk 12]
```diff
@@ -1,235 +1,417 @@
...
+        let r = unsafe { h.as_ref() };
+        r.squawk();
```
Comment (nicklan) [inline ffi/src/handle.rs:401]:
```suggestion
        assert_eq!(r.squawk(), "Bar!");
```

File: ffi/src/handle.rs [hunk 13]
```diff
@@ -1,235 +1,417 @@
...
-        let ptr = handle as *const Self::Target;
-        let _ = unsafe { Arc::from_raw(ptr) };
+        let b = Bar {
+            x: 10,
+            y: "hello".into(),
+        };
+        let t: Arc<dyn Baz> = Arc::new(b);
+        let h: Handle<SharedBaz> = t.into();
+        let r = unsafe { h.as_ref() };
+        r.squawk();
```
Comment (nicklan) [inline ffi/src/handle.rs:413]:
```suggestion
        assert_eq!(r.squawk(), "Bar!");
```

File: ffi/src/handle.rs [hunk 14]
```diff
@@ -1,235 +1,417 @@
...
+        let r = unsafe { h.as_arc() };
+        r.squawk();
```
Comment (nicklan) [inline ffi/src/handle.rs:415]:
```suggestion
        assert_eq!(r.squawk(), "Bar!");
```

**Review by nicklan — APPROVED** [review]
Yeah, overall this is really cool!

Left a few suggestions for cleanup, but lgtm

File: ffi/src/scan.rs [hunk 15]
```diff
@@ -83,27 +100,34 @@ unsafe fn get_raw_arrow_data_impl(
     Ok(Box::leak(ret_data))
 }
 
-impl BoxHandle for Scan {}
+pub struct ScanHandle {
```
Comment (nicklan) [inline ffi/src/scan.rs:103]: Probably should name this `SharedScan` so we end up with a `HandleSharedScan` on the c side and not a `HandleScanHandle`

File: ffi/src/scan.rs [hunk 16]
```diff
@@ -112,27 +136,34 @@ unsafe fn scan_impl(
             scan_builder = scan_builder.with_predicate(predicate);
         }
     }
-    Ok(BoxHandle::into_handle(scan_builder.build()))
+    Ok(Box::new(scan_builder.build()).into())
 }
 
-impl BoxHandle for GlobalScanState {}
+pub struct GlobalScanStateHandle {
```
Comment (nicklan) [inline ffi/src/scan.rs:142]:
```suggestion
pub struct MutGlobalScanState {
```

Not sure what prefix we want to use instead of "Shared" for `Mutable = True` things. Maybe `Mut`, or maybe `Exclusive`?

File: ffi/src/scan.rs (hunk 15 above)
Comment (nicklan) [reply ffi/src/scan.rs:103]: Actually it's not shared I guess, so either `MutScan` or `ExclusiveScan`?

File: ffi/src/scan.rs [hunk 17]
```diff
@@ -147,7 +178,14 @@ pub struct KernelScanDataIterator {
     engine: Arc<dyn ExternEngine>,
 }
 
-impl BoxHandle for KernelScanDataIterator {}
+pub struct KernelScanDataIteratorHandle {
```
Comment (nicklan) [inline ffi/src/scan.rs:181]: remove Handle from name

File: ffi-proc-macros/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+                _ => {
+                    return Err(Error::new(ident.span(), "unknown or duplicated field"));
+                }
+            }
+        }
+        let mutable = mutable.ok_or_else(|| input.error("missing `mutable` field"))?;
+        if sized.is_none() && is_trait {
+            sized = Some(false);
+        }
+        let sized = sized.ok_or_else(|| input.error("missing `sized` field"))?;
```
Comment (nicklan) [inline ffi-proc-macros/src/lib.rs:65]: can't we just `unwrap` here, we just ensured it's not `None` above

**Review by nicklan — CHANGES_REQUESTED** [review]
quick flush, more to come

File: ffi/src/lib.rs [hunk 19]
```diff
@@ -14,9 +14,10 @@ use delta_kernel::expressions::{BinaryOperator, Expression, Scalar};
 use delta_kernel::schema::{ArrayType, DataType, MapType, PrimitiveType, StructType};
 use delta_kernel::snapshot::Snapshot;
 use delta_kernel::{DeltaResult, Engine, Error};
+use delta_kernel_ffi_macros::handle_descriptor;
 
-mod handle;
-use handle::{ArcHandle, SizedArcHandle, Unconstructable};
+pub mod handle;
```
Comment (nicklan) [inline ffi/src/lib.rs:19]: Do we expect things outside of the ffi crate to implement `Handles`? If not, I think we can `pub(crate)` this

File: ffi/src/scan.rs [hunk 20]
```diff
@@ -7,45 +7,54 @@ use std::sync::Arc;
...
+    let _data = unsafe { data.as_mut() };
+    todo!() // See TODO comment for EngineDataHandle
 }
 
 /// Struct to allow binding to the arrow [C Data
 /// Interface](https://arrow.apache.org/docs/format/CDataInterface.html). This includes the data and
 /// the schema.
 #[cfg(feature = "default-engine")]
-#[repr(C)]
+// FIXME: cbindgen doesn't recognize the FFI_ArrowFoo fields in this struct
```
Comment (nicklan) [inline ffi/src/scan.rs:53]: I fixed this here: https://github.com/delta-incubator/delta-kernel-rs/pull/203/files#diff-e4dc56c14dad42367f172aee9d547df5959ef6b1b3a4575dfc4d989c0f97ab8bR21

File: ffi/src/scan.rs [hunk 21]
```diff
@@ -56,19 +65,20 @@ pub struct ArrowFFIData {
...
 }
 
 #[cfg(feature = "default-engine")]
 unsafe fn get_raw_arrow_data_impl(
-    data_handle: *mut EngineDataHandle,
+    data: Handle<EngineDataHandle>,
 ) -> DeltaResult<*mut ArrowFFIData> {
-    let boxed_data = unsafe { Box::from_raw(data_handle) };
-    let data = boxed_data.data;
+    // TODO(frj): This consumes the handle. Is that what we really want?
```
Comment (nicklan) [inline ffi/src/scan.rs:80]: I believe so yet. afaict this transfers ownership of the data to the engine. It's a little unclear from the arrow docs, but if I try and free the data after using this I get a segfault so I'm pretty sure

File: kernel/Cargo.toml [hunk 22]
```diff
@@ -70,6 +70,7 @@ default-engine = [
   "parquet/object_store",
   "reqwest",
   "tokio",
+  "tokio/sync"
```
Comment (nicklan) [inline kernel/Cargo.toml:73]: why was this needed?

File: ffi/src/lib.rs (hunk 6 above)
Comment (scovich) [reply ffi/src/lib.rs]: Good point. Deleted this and a bunch of other junk

File: ffi/src/scan.rs (hunk 15 above)
Comment (scovich) [reply ffi/src/scan.rs]: Yeah scan is a weird one (duckdb was hitting concurrency bugs here). It has to be mutable, because consuming from the scan modifies its internal state. I think we fixed the bug by adding a mutex internally, but technically duckdb should probably be doing the mutex work to ensure there's only one mutable reference possible at a time??

File: ffi/src/scan.rs [hunk 23]
```diff
@@ -83,27 +93,28 @@ unsafe fn get_raw_arrow_data_impl(
     Ok(Box::leak(ret_data))
 }
 
-impl BoxHandle for Scan {}
+#[handle_descriptor(target=Scan, mutable=true, sized=true)]
+pub struct ScanHandle;
```
Comment (scovich) [inline ffi/src/scan.rs:97]: Rescuing https://github.com/delta-incubator/delta-kernel-rs/pull/182#discussion_r1603862042
> Actually it's not shared I guess, so either `MutScan` or `ExclusiveScan`?

Yeah scan is a weird one (duckdb was hitting concurrency bugs here). It has to be mutable, because consuming from the scan modifies its internal state. I think we fixed the bug by adding a mutex internally, but technically duckdb should probably be doing the mutex work to ensure there's only one mutable reference possible at a time??

File: ffi/src/scan.rs [hunk 24]
```diff
@@ -112,27 +123,28 @@ unsafe fn scan_impl(
             scan_builder = scan_builder.with_predicate(predicate);
         }
     }
-    Ok(BoxHandle::into_handle(scan_builder.build()))
+    Ok(Box::new(scan_builder.build()).into())
 }
 
-impl BoxHandle for GlobalScanState {}
+#[handle_descriptor(target=GlobalScanState, mutable=true, sized=true)]
+pub struct GlobalScanStateHandle;
```
Comment (scovich) [inline ffi/src/scan.rs:130]: Rescuing https://github.com/delta-incubator/delta-kernel-rs/pull/182#discussion_r1603863446
> Not sure what prefix we want to use instead of "Shared" for `Mutable = True` things. Maybe `Mut`, or maybe `Exclusive`?

Shouldn't we just have the rule that it's mutable unless clearly stated otherwise? That kind of "opt out" is safer than an "opt in" approach?

File: ffi/src/handle.rs [hunk 25]
```diff
@@ -1,235 +1,527 @@
...
 
-// Make it easy to use (the module was only used to enforce member privacy).
-pub type Unconstructable = unconstructable::Unconstructable;
+    /// Clones this handle, increasing the underlying object's refcount by one. The new handle is
+    /// independent of the source handle, and should be dropped when it is no longer needed.
+    ///
+    /// # Safety
+    ///
+    /// Caller asserts that the handle is [valid][Handle#Validity].
+    unsafe fn clone_handle(&self) -> Self;
```
Comment (nicklan) [inline ffi/src/handle.rs:57]: This isn't used. Can we call it `_clone_handle` or `#[allow(dead_code)]` it

File: ffi/src/scan.rs (hunk 24 above)
Comment (nicklan) [reply ffi/src/scan.rs:130]: Yeah, for this one it's probably fine because we can use `GlobalScanStateHandle*` and don't have to use `HandleGlobalScanStateHandle`.

For some of the `ExternResult` stuff the types get... not so nice (i.e. `ExternResultHandleKernelScanDataIteratorHandle`)

Comment (scovich) [reply ffi/src/scan.rs:130]: As in, if it doesn't say "shared" in the name... it's not shared.

File: ffi/src/scan.rs [hunk 26]
```diff
@@ -147,7 +159,8 @@ pub struct KernelScanDataIterator {
     engine: Arc<dyn ExternEngine>,
 }
 
-impl BoxHandle for KernelScanDataIterator {}
+#[handle_descriptor(target=KernelScanDataIterator, mutable=true, sized=true)]
+pub struct KernelScanDataIteratorHandle;
```
Comment (scovich) [inline ffi/src/scan.rs:163]: Rescuing https://github.com/delta-incubator/delta-kernel-rs/pull/182#discussion_r1603864171

> remove Handle from name

This is a conflict between the current cbindgen capability vs. what cbindgen "should" be, combined with name mangling for C:
* Today -- cbindgen exposes `Handle<T>` in C++ as `template <typename T> typedef Handle<T> = T`, but any given `Handle<FooHandle>` will mangle to `HandleFooHandle` in the resulting C header. Which is indeed annoying. 
* Some day (ideally), cbindgen would understand that `Handle<T>` is really just a `T*` and would expose it as such. The resulting header would just have `FooHandle*` everywhere.

However, even today both C and C++ can still use a simple `FooHandle*` -- and probably _should_ do so, since the typedef doesn't add anything useful (we already have the indirection and opacity we desire from the handle itself). The one exception is things like `ExternResult<Handle<FooHandle>>` which with today's cbindgen would continue mangling badly. And since it's not a typedef we can't just ignore it (have to actually fix cbindgen). 

I guess the other approach we could do is give the kernel types the same names as their corresponding FFI types? Something like:
```rust
use delta_kernel::{Snapshot => KernelSnapshot};

#[handle_descriptor(target=KernelSnapshot, mutable=true, sized=true)]
pub struct Snapshot;
```
Downside is, that only has type safety on the rust side; on the C/C++ side one could be tempted to cast `delta_kernel_ffi::Snapshot` as `delta_kernel::Snapshot`, since they're the same name. 

Then again, if somebody's really determined to do shady casting of a type we've officially defined as opaque, then they probably deserve whatever bugs they cause.

File: ffi-proc-macros/src/lib.rs [hunk 27]
```diff
@@ -0,0 +1,124 @@
...
+        })
+        | Fields::Unnamed(FieldsUnnamed {
+            unnamed: ref fields,
+            ..
+        }) if !fields.is_empty() => {
+            return Error::new(fields.span(), "A handle struct must not define any fields")
+                .into_compile_error()
+                .into();
+        }
+        _ => {}
```
[the remark this answers is not in the document]
Comment (scovich) [reply ffi-proc-macros/src/lib.rs:101]: No, there's an `if` on the other case(s)...

File: ffi-proc-macros/src/lib.rs [hunk 28]
```diff
@@ -0,0 +1,124 @@
...
+                }
+                "sized" if sized.is_none() => {
+                    if value.value() && is_trait {
+                        return Err(Error::new(value.span(), "trait objects are never Sized"));
+                    }
+                    sized = Some(value.value());
+                }
+                _ => {
+                    return Err(Error::new(ident.span(), "unknown or duplicated field"));
+                }
```
[the remark this answers is not in the document]
Comment (scovich) [reply ffi-proc-macros/src/lib.rs:58]: I originally didn't think it was necessary given the span is available? 
```
error: unknown or duplicated field
   --> ffi/src/scan.rs:258:50
    |
258 | #[handle_descriptor(target=DvInfo, mutable=true, blah=true, sized=true)]
    |                                                  ^^^^
```
(but I could add it if you prefer)

File: ffi-proc-macros/src/lib.rs (hunk 18 above)
Comment (scovich) [reply ffi-proc-macros/src/lib.rs:65]: L63 only applies to trait targets (for which `sized` is optional); struct targets still require it, since I couldn't figure out a way to tell at macro expansion time whether the target struct is `Sized` or not...

File: ffi/src/lib.rs (hunk 19 above)
Comment (scovich) [reply ffi/src/lib.rs:19]: No. Good point!

Comment (scovich) [reply ffi/src/lib.rs:19]: Hmm, making it `pub(crate)` suppresses docs for it tho... seems like it would be nice to have docs even for internal classes, for benefit of kernel developers?

File: ffi/src/scan.rs (hunk 26 above)
Comment (nicklan) [reply ffi/src/scan.rs:163]: Yeah, I think this is okay for now. I prefer to keep the ffi types distinct from the kernel types, just to avoid confusion.

Changes to cbindgen that would be nice:
- rename on a generic struct renames on the mangled name. So a rename rule of `ExternResult` -> `Result` would let us have things like `ResultFooHandle`, since it's not really "extern" for the engine
- configure the mangling to put the outer type at the end, so we could get `FooHandleResult` which is probably the most "correct" name

File: ffi/src/scan.rs (hunk 21 above)
Comment (scovich) [reply ffi/src/scan.rs:80]: Yes, today the code transfers ownership to the engine... but _should_ it? Do we _want_ to be allocating that data over and over? Or allocating once and rewriting it over and over?

File: kernel/Cargo.toml (hunk 22 above)
Comment (scovich) [reply kernel/Cargo.toml:73]: Bad merge, my branch still had the first attempt at fixing the concurrency bug, which used that package.

File: ffi/src/handle.rs (hunk 25 above)
Comment (scovich) [reply ffi/src/handle.rs:57]: I'm a bit confused actually -- this is a `pub trait` whose methods therefore are all public. 
Why would it even trigger dead code warnings in the first place?

File: ffi-proc-macros/src/lib.rs (hunk 27 above)
[the remark this answers is not in the document]
Comment (nicklan) [reply ffi-proc-macros/src/lib.rs:101]: ah right.

File: ffi/src/lib.rs (hunk 19 above)
Comment (nicklan) [reply ffi/src/lib.rs:19]: well, if you add `--document-private-items` to `cargo doc` it will generate those docs.

not _ideal_ for people who want to hack on kernel, but we could probably document that.

File: ffi-proc-macros/src/lib.rs (hunk 28 above)
[the remark this answers is not in the document]
Comment (nicklan) [reply ffi-proc-macros/src/lib.rs:58]: Ahh, okay I hadn't actually seen the outputted span. This is likely fine as is then

File: ffi-proc-macros/src/lib.rs (hunk 18 above)
Comment (nicklan) [reply ffi-proc-macros/src/lib.rs:65]: Ahh right, and I guess we could maybe have zero sized structs (although unlikely). Okay, makes sense then.

File: ffi/src/handle.rs (hunk 25 above)
Comment (nicklan) [reply ffi/src/handle.rs:57]: Sorry, that'll happen if you make the other change I suggested where handle is `pub(crate) mod handle;`

File: ffi/src/scan.rs (hunk 21 above)
Comment (nicklan) [reply ffi/src/scan.rs:80]: Well, in this case we want to transfer actual raw arrow data over. So either we have to `clone` it and keep the previous one alive, or consume it. 

From my comments in the code below , I wasn't quite clear if `let array = arrow_data::ffi::FFI_ArrowArray::new(&array_data);` consumed the data, but afaict, it does.

So for efficiency's sake I think we'd want to keep the semantics that this destroys passed in data and you get back the raw arrow version of it.

File: ffi/src/scan.rs (hunk 23 above)
Comment (scovich) [reply ffi/src/scan.rs:97]: Thinking more about this -- we should be doing an `Arc<Mutex<T>>` for this case, instead of locking internally inside the iterator. Right now it may "work" but we're violating the mutable borrow invariant. Unfortunately, the existing `Mutex` seems needed in order for the result of `new_async_read_iterator` to be `Send`:
```
error[E0277]: `std::sync::mpsc::Receiver<std::result::Result<arrow_array::RecordBatch, error::Error>>` cannot be shared between threads safely
   --> kernel/src/engine/default/file_stream.rs:151:12
    |
151 |         Ok(Box::new(it))
    |            ^^^^^^^^^^^^ `std::sync::mpsc::Receiver<std::result::Result<arrow_array::RecordBatch, error::Error>>` cannot be shared between threads safely
    |
    = help: within `std::iter::Map<std::sync::mpsc::IntoIter<std::result::Result<arrow_array::RecordBatch, error::Error>>, {closure@kernel/src/engine/default/file_stream.rs:145:18: 145:23}>`, the trait `std::marker::Sync` is not implemented for `std:\
:sync::mpsc::Receiver<std::result::Result<arrow_array::RecordBatch, error::Error>>`
note: required because it appears within the type `IntoIter<Result<RecordBatch, Error>>`
   --> /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/sync/mpsc/mod.rs:302:12
note: required because it appears within the type `Map<IntoIter<Result<RecordBatch, Error>>, {closure@file_stream.rs:145:18}>`
   --> /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/core/src/iter/adapters/map.rs:61:12
    = note: required for the cast from `Box<Map<IntoIter<Result<RecordBatch, Error>>, {closure@file_stream.rs:145:18}>>` to `Box<dyn Iterator<Item = Result<Box<dyn EngineData>, Error>> + Send + Sync>`
```

**Review by nicklan — APPROVED** [review]
lgtm!

File: ffi-proc-macros/src/lib.rs (hunk 28 above)
[the remark this answers is not in the document]
Comment (scovich) [reply ffi-proc-macros/src/lib.rs]: Decided to add it anyway (wasn't difficult, and most compiler errors do it)

File: ffi/src/scan.rs (hunk 26 above)
Comment (scovich) [reply ffi/src/scan.rs]: As a side effect of fixing https://github.com/delta-incubator/delta-kernel-rs/pull/182#discussion_r1604079239, all handles became shared except the one for `EngineData`, and so could be renamed as `SharedXxx`. The current code for engine data handles is illegal/incorrect, and will likely change, so for now I just imported the kernel `EngineData` as `KernelEngineData` so the FFI one could be called `EngineData`.

File: kernel/src/engine/default/file_stream.rs [hunk 29]
```diff
@@ -140,14 +140,10 @@ impl FileStream {
         // underlying stream. This should not be a bottleneck because the stream will immediately
         // return an item (if ready) and would anyway block if not ready. Additionally, each
         // iterator element corresponds to a file access whose cost dwarfs any mutex overhead.
-        let it = receiver
+        let mut it = receiver
             .into_iter()
             .map(|rbr| rbr.map(|rb| Box::new(ArrowEngineData::new(rb)) as _));
-        let it = std::sync::Mutex::new(it);
```
Comment (scovich) [inline kernel/src/engine/default/file_stream.rs:146]: This is too deep and thus too late; moved to top-level FFI code instead.

File: ffi/src/handle.rs [hunk 30]
```diff
@@ -1,235 +1,616 @@
...
-///
-/// * The `Target` type is `Sync`
-///
-/// * The handle is (correctly) `Send`
-pub trait ArcHandle: Sized {
-    /// The target type this handle represents.
-    type Target: ?Sized + Send + Sync;
+pub trait HandleDescriptor {
+    /// The true type of the handle's underlying object
+    type Target: ?Sized + Send;
```
Comment (scovich) [inline ffi/src/handle.rs:37]: NOTE: Mutable handles only need to be `Send` (not `Sync`), because any use by external code that would rely on them being `Sync` would actually break the `Send` contract: Mutable handles are mutable references, and because handles are opaque to extern code, rust FFI code always receives the handle as mutable. Doing that from multiple threads means multiple mutable reference now exist in rust at the same time, which produces undefined behavior.

File: ffi/src/handle.rs [hunk 31]
```diff
@@ -1,235 +1,616 @@
...
+    where
+        T: ?Sized,
+        H: HandleDescriptor<Target = T, Mutable = True, Sized = S>,
+    {
+    }
+
+    pub trait SharedHandle {}
+    impl<T, S, H> SharedHandle for H
+    where
+        T: ?Sized + Sync,
```
Comment (scovich) [inline ffi/src/handle.rs:244]: All handles are `Send`, but shared handles also need to be `Sync` because the engine is free to pass them as args to kernel functions from any number of threads at the same time.

File: ffi/src/handle.rs [hunk 32]
```diff
@@ -1,235 +1,616 @@
...
+    /// this requirement applies even for API calls that appear to be read-only (because Rust code
+    /// always receives the handle as mutable).
+    ///
+    /// NOTE: Because the underlying type is always [`Sync`], multi-threaded external code can
+    /// freely access shared (non-mutable) handles.
+    ///
+    /// cbindgen:transparent-typedef
+    #[repr(transparent)]
+    pub struct Handle<H: HandleDescriptor> {
+        ptr: NonNull<H>,
```
Comment (scovich) [inline ffi/src/handle.rs:80]: TODO: Make the shared handles produce `const T*` on C/C++ side.

Comment (scovich) [reply ffi/src/handle.rs:80]: Update: `NonNull<T>` always produces `T*` (mutable) pointers. Nothing we can do about that, unfortunately.

File: ffi/src/handle.rs [hunk 33]
```diff
@@ -1,235 +1,616 @@
...
+        let b = Bar {
+            x: rand::random::<usize>(),
+            y: rand::random::<usize>().to_string(),
+        };
+        let s = format!("{b:?}");
+        let h: Handle<SharedBar> = Arc::new(b).into();
+        let r = unsafe { h.as_ref() };
+        assert_eq!(format!("{r:?}"), s);
+
+        // error[E0599]: the method `as_mut` exists for struct `Handle<BarHandle>`,
```
[the remark this answers is not in the document]
Comment (scovich) [reply ffi/src/handle.rs:564]: I think the bigger issue is that:
> [doc tests] link against only the public items of your crate; if you need to test private items, you need to write a unit test

File: kernel/src/engine/default/file_stream.rs [hunk 34]
```diff
@@ -140,14 +140,10 @@ impl FileStream {
...
-        let it = std::iter::from_fn(move || match it.lock() {
-            Ok(mut i) => i.next(),
-            Err(_) => Some(Err(Error::generic("Poisoned mutex"))),
-        });
+        let it = std::iter::from_fn(move || it.next());
```
Comment (roeap) [inline kernel/src/engine/default/file_stream.rs:146]: Do we need the additional `std::iter:from_fn(...)`? I think we could simplify to

```rust
let it = receiver
    .into_iter()
    .map(|rbr| rbr.map(|rb| Box::new(ArrowEngineData::new(rb)) as _));
Ok(Box::new(it))
```

The comment above may also be no longer relevant?

**Review by roeap — APPROVED** [review]
👍

I especially like the docs - to me it seems there is a clean and digestible story what engines have to look out for.

**Review by hntd187 — APPROVED** [review]
Read through twice, looks solid. We'll have to setup fuzzing on the C side for some additional hardening but fantastic docs and I think this establishes a good contract.

File: ffi/src/handle.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (scovich) [reply ffi/src/handle.rs]: (found a way to avoid `unimplemented!` altogether)
